### PR TITLE
Characters API Client

### PR DESCRIPTION
**Pull request by pedrovgs**
Added all the code needed to be able to retrieve characters from the Marvel API. Pending error handling  which is going to be the next request.

### CODE REVIEW

File: MarvelApiClient/CharactersParser.swift [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+//  MarvelApiClient
+//
+//  Created by Pedro Vicente Gomez on 16/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+import SwiftyJSON
+
+class CharactersParser : Parser {
```
Comment (pedrovgs) [inline MarvelApiClient/CharactersParser.swift]: Thinking in remove the Parser protocol. What do you think guys?

File: MarvelApiClientTests/CharactersApiClientTests.swift [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+        }
+    }
+
+    private func beGetCharacterByIdSuccess<T>() -> MatcherFunc<T?> {
+        return MatcherFunc { actualExpression, failureMessage in
+            failureMessage.postfixMessage = "be success"
+            let future = try actualExpression.evaluate() as! Future<CharacterDTO,NSError>
+            return future.isSuccess
+        }
+    }
```
Comment (pedrovgs) [inline MarvelApiClientTests/CharactersApiClientTests.swift]: @davideme can you help me to refactor these two methods to be the same waiting for a future success?

File: MarvelApiClient/CharactersApiClient.swift [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+import Foundation
+import BrightFutures
+
+public class CharactersApiClient : BaseApiClient {
+
+    private let parser: CharactersParser
+
+    init(baseEndpoint: String, timeProvider: TimeProvider, httpClient: HttpClient, parser: CharactersParser) {
+        self.parser = parser
+        super.init(baseEndpoint: baseEndpoint, timeProvider: timeProvider, httpClient: httpClient)
```
Comment (davideme) [inline MarvelApiClient/CharactersApiClient.swift]: Wouldn't it be better to have BaseApiClient as a property instead of subclassing ?
If the client use it with dependency injection it could reuse the same instance accros [Chararactes|Comics|Authros]ApiClient.

File: MarvelApiClient/CharactersApiClient.swift [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+//  MarvelApiClient
+//
+//  Created by Pedro Vicente on 14/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+import BrightFutures
+
+public class CharactersApiClient : BaseApiClient {
```
Comment (davideme) [inline MarvelApiClient/CharactersApiClient.swift]: If we use getAll getById as method name, I prefer that we use the CharactersApiDatasource

File: MarvelApiClient/Parser.swift [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+//  Parser.swift
+//  MarvelApiClient
+//
+//  Created by Pedro Vicente Gomez on 16/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+
+protocol Parser {
```
Comment (davideme) [inline MarvelApiClient/Parser.swift]: Could we have a method that return a list and another one 1 item ?

File: MarvelApiClient/HttpClient.swift [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+//  Created by Pedro Vicente Gomez on 12/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+import BrightFutures
+
+protocol HttpClient {
+
+    func send(httpRequest: HttpRequest) -> Future<HttpResponse, NSError>
```
Comment (davideme) [inline MarvelApiClient/HttpClient.swift]: Would be better if instead of NSError, we use a Generic subclassing ErrorType, to be future proof.

File: MarvelApiClient/CharactersApiClient.swift [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+
+    private let parser: CharactersParser
+
+    init(baseEndpoint: String, timeProvider: TimeProvider, httpClient: HttpClient, parser: CharactersParser) {
+        self.parser = parser
+        super.init(baseEndpoint: baseEndpoint, timeProvider: timeProvider, httpClient: httpClient)
+    }
+    
+    public func getAll(offset: Int, limit: Int) -> Future<GetCharactersDTO,NSError> {
+        let params = [MarvelApiParams.Offset : "\(offset)", MarvelApiParams.Limit : "\(limit)"]
```
Comment (davideme) [inline MarvelApiClient/CharactersApiClient.swift]: Does negative value of offset and int are authorized ?
Should we filter them in the client or let them fail in server ?

File: MarvelApiClient/BaseApiClient.swift [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+//  MarvelApiClient
+//
+//  Created by Pedro Vicente on 14/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+import BrightFutures
+
+public class BaseApiClient {
```
Comment (davideme) [inline MarvelApiClient/BaseApiClient.swift]: If it's public let's prefix it with Marvel.
MarvelBaseApiClient.

File: MarvelApiClient/MarvelHashGenerator.swift [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+//  Created by Pedro Vicente on 14/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+import CryptoSwift
+
+class MarvelHashGenerator {
+
+    static func generateHash(timestamp: Int, privateKey: String, publicKey: String) -> String {
```
Comment (davideme) [inline MarvelApiClient/MarvelHashGenerator.swift]: Why is not part of the BaseClientApi ?

File: MarvelApiClient/TimeProvider.swift [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+//
+//  Created by Pedro Vicente on 14/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+
+class TimeProvider {
+
+    func currentTimeMillis() -> Int {
```
Comment (davideme) [inline MarvelApiClient/TimeProvider.swift]: Let's move it as an extension of NSDate.

**Comment by davideme** [comment]
We will use the convention that Abbreviations in UpperCase.

File: MarvelApiClient/CharactersApiClient.swift (hunk 4 above)
Comment (pedrovgs) [inline MarvelApiClient/CharactersApiClient.swift]: In this context this is not related to a Repository patterns. I prefer to change the method names before to use DataSource as suffix. Any suggestion for the method names?

File: MarvelApiClient/Parser.swift (hunk 5 above)
Comment (pedrovgs) [inline MarvelApiClient/Parser.swift]: Sure, but we don't need it now and the implementation will be empty. I can add it if you want :)

File: MarvelApiClient/HttpClient.swift (hunk 6 above)
Comment (pedrovgs) [inline MarvelApiClient/HttpClient.swift]: That's what I mentioned in the PR description. My next task is going to be to implement the error handling.

File: MarvelApiClient/CharactersApiClient.swift (hunk 7 above)
Comment (pedrovgs) [inline MarvelApiClient/CharactersApiClient.swift]: Fail in server could be enough for this project :)

File: MarvelApiClient/MarvelHashGenerator.swift (hunk 9 above)
Comment (pedrovgs) [inline MarvelApiClient/MarvelHashGenerator.swift]: It is. ``MarvelBaseApiClient line 38``

File: MarvelApiClient/TimeProvider.swift (hunk 10 above)
Comment (pedrovgs) [inline MarvelApiClient/TimeProvider.swift]: I can't . Because I need to be able to mock the time :(

File: MarvelApiClient/CharactersApiClient.swift (hunk 3 above)
Comment (pedrovgs) [inline MarvelApiClient/CharactersApiClient.swift]: Done! :)

File: MarvelApiClient/CharactersApiClient.swift (hunk 7 above)
Comment (pedrovgs) [inline MarvelApiClient/CharactersApiClient.swift]: I've added an assert :)

File: MarvelApiClient/BaseApiClient.swift (hunk 8 above)
Comment (pedrovgs) [inline MarvelApiClient/BaseApiClient.swift]: Done! :)

**Comment by pedrovgs** [comment]
@davideme I've applied all your suggestions including the one to move the naming convention to upper case always :) Let me know if you have any extra suggestion.

**Comment by pedrovgs** [comment]
@davideme can we merge this one and the previous one?

File: MarvelApiClient/HTTPVerb.swift [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+//  HttpVerb.swift
+//  MarvelApiClient
+//
+//  Created by Pedro Vicente Gomez on 12/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+
+enum HTTPVerb {
```
Comment (davideme) [inline MarvelApiClient/HTTPVerb.swift]: The name should be HTTPMethod based on the rfc http://www.w3.org/Protocols/rfc2616/rfc2616-sec9.html
And if HTTPMethod extend String (HTTPMethod: String)you would have a direct conversion to the raw value.
HTTPMethod.GET.rawValue == "GET"

Also for BothamNetworking we should list all Method available in the standard.
OPTIONS GET HEAD POST PUT DELETE TRACE CONNECT

File: MarvelApiClient/AlamofireHTTPClient.swift [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+                return promise.success(response)
+            }
+        }
+        return promise.future
+    }
+}
+
+private extension HTTPVerb {
+
+    private func toAlamofireVerb() -> Alamofire_Result.Method {
```
Comment (davideme) [inline MarvelApiClient/AlamofireHTTPClient.swift]: Could this be an extension of Alamofire_Result.Method instead, the actual signature feel a bit redundant to me.

File: MarvelApiClient/HTTPRequest.swift [hunk 13]
```diff
@@ -0,0 +1,18 @@
...
+//  Created by Pedro Vicente Gomez on 12/11/15.
+//  Copyright © 2015 GoKarumi S.L. All rights reserved.
+//
+
+import Foundation
+
+struct HTTPRequest {
+
+    let url: String
+    let parameters: [String : String]?
```
Comment (davideme) [inline MarvelApiClient/HTTPRequest.swift]: instead of URL and Parameters we usually use NSURL in iOS.

File: MarvelApiClient/MarvelBaseApiClient.swift [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+
+    init(baseEndpoint: String, timeProvider: TimeProvider, httpClient: HTTPClient) {
+        self.baseEndpoint = baseEndpoint
+        self.timeProvider = timeProvider
+        self.httpClient = httpClient
+    }
+
+    func sendRequest(verb: HTTPVerb, path: String,
+                     params: [String:String]? = [String:String]()) -> Future<HTTPResponse, NSError> {
+        let parameters = addDefaultParams(params)
```
Comment (davideme) [inline MarvelApiClient/MarvelBaseApiClient.swift]: And if we user NSURL, we use NSURLComponents tu build the string and and NSURL with it.

File: MarvelApiClient/Parser.swift (hunk 5 above)
Comment (davideme) [inline MarvelApiClient/Parser.swift]: Ok let's move this idea at BothamNetworking.

File: MarvelApiClient/TimeProvider.swift (hunk 10 above)
Comment (davideme) [inline MarvelApiClient/TimeProvider.swift]: :ok_hand:

File: MarvelApiClient/HTTPRequest.swift [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+    let headers: [String : String]?
+    let verb: HTTPVerb
+
```
Comment (davideme) [inline MarvelApiClient/HTTPRequest.swift]: Could we add a let HTTPBody: NSData? to allow post method ?

File: MarvelApiClient/MarvelHashGenerator.swift (hunk 9 above)
Comment (davideme) [inline MarvelApiClient/MarvelHashGenerator.swift]: :ok:

File: MarvelApiClient/MarvelBaseApiClient.swift (hunk 14 above)
Comment (pedrovgs) [inline MarvelApiClient/MarvelBaseApiClient.swift]: As we've discussed we are going to implement this in a next PR.

File: MarvelApiClient/HTTPRequest.swift (hunk 13 above)
Comment (pedrovgs) [inline MarvelApiClient/HTTPRequest.swift]: As we've discussed we are going to implement this in a next PR.

File: MarvelApiClient/AlamofireHTTPClient.swift (hunk 12 above)
Comment (pedrovgs) [inline MarvelApiClient/AlamofireHTTPClient.swift]: With the usage of an enum extending from String this method is not needed anymore.

File: MarvelApiClient/HTTPVerb.swift (hunk 11 above)
Comment (pedrovgs) [inline MarvelApiClient/HTTPVerb.swift]: Done :)

**Comment by davideme** [comment]
:+1:

**Comment by pedrovgs** [comment]
Thanks for all your feedbak!! I'm going to merge this PR and continue working on the core components for BothamNetworking :)
